ddtrace/tracer/civisibility_transport.go

```diff
@@ -0,0 +1,194 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+package tracer
+
+import (
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/civisibility/constants"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/version"
+)
+
+// Constants for CI Visibility API paths and subdomains.
+const (
+	TestCycleSubdomain = "citestcycle-intake" // Subdomain for test cycle intake.
+	TestCyclePath      = "api/v2/citestcycle" // API path for test cycle.
+	EvpProxyPath       = "evp_proxy/v2"       // Path for EVP proxy.
+)
+
+// Ensure that civisibilityTransport implements the transport interface.
+var _ transport = (*civisibilityTransport)(nil)
+
+// civisibilityTransport is a structure that handles sending CI Visibility payloads
+// to the Datadog endpoint, either in agentless mode or through the EVP proxy.
+type civisibilityTransport struct {
+	config           *config           // Configuration for the tracer.
+	testCycleURLPath string            // URL path for the test cycle endpoint.
+	headers          map[string]string // HTTP headers to be included in the requests.
+	agentless        bool              // Gets if the transport is configured in agentless mode (eg: Gzip support)
+}
+
+// newCiVisibilityTransport creates and initializes a new civisibilityTransport
+// based on the provided tracer configuration. It sets up the appropriate headers
+// and determines the URL path based on whether agentless mode is enabled.
+//
+// Parameters:
+//
+//	config - The tracer configuration.
+//
+// Returns:
+//
+//	A pointer to an initialized civisibilityTransport instance.
+func newCiVisibilityTransport(config *config) *civisibilityTransport {
+	// Initialize the default headers with encoder metadata.
+	defaultHeaders := map[string]string{
+		"Datadog-Meta-Lang":             "go",
+		"Datadog-Meta-Lang-Version":     strings.TrimPrefix(runtime.Version(), "go"),
+		"Datadog-Meta-Lang-Interpreter": runtime.Compiler + "-" + runtime.GOARCH + "-" + runtime.GOOS,
+		"Datadog-Meta-Tracer-Version":   version.Tag,
+		"Content-Type":                  "application/msgpack",
+	}
+	if cid := internal.ContainerID(); cid != "" {
+		defaultHeaders["Datadog-Container-ID"] = cid
+	}
+	if eid := internal.EntityID(); eid != "" {
+		defaultHeaders["Datadog-Entity-ID"] = eid
+	}
+
+	// Determine if agentless mode is enabled through an environment variable.
+	agentlessEnabled := internal.BoolEnv(constants.CiVisibilityAgentlessEnabledEnvironmentVariable, false)
+
+	testCycleURL := ""
+	if agentlessEnabled {
+		// Agentless mode is enabled.
+		defaultHeaders["dd-api-key"] = os.Getenv(constants.APIKeyEnvironmentVariable)
```

Comment:
Do we need to notify the user in some way if they haven't set an API key env var?

Reply:
you're right, thanks for the catch.

Reply:
done in [5862d18](https://github.com/DataDog/dd-trace-go/pull/2739/commits/5862d18e6a1c93770fabb8f38009059103edab5b)